### ISSUE TITLE
[GCP] Document how to use Kubeflow with shared VPC

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#3082]
We need to document how to use Kubeflow with shared VPC.
https://cloud.google.com/vpc/docs/shared-vpc

**Comment by jlewi** [comment]
Once #3124 is submitted we need to update the docs. 

We will need to specify the config or version to use and verify that it works.

**Comment by jlewi** [comment]
Here was an initial version of docs
https://www.kubeflow.org/docs/gke/private-clusters/#enable-vpc-service-controls-in-your-project

The information is outdated. We will need to update the information on how to replicate the images to private GCR.

The good news is @xaniasd just submitted #4222 which allows bulk replacement of the images.

**Comment by garrettwong** [comment]
Do we have any examples/guidance of leveraging a Shared VPC network (one that does not belong to the current GCP project)?  

We're adhering to Principle of Least Permission and replacement of the `default` network parameter the Deployment Manager scripts that I could find with (below format) does not seem to work.  

- projects/${SHARED_VPC_PROJECT_ID}/global/networks/${NETWORK}

**Comment by jrabiah** [comment]
I was able to get this working by running through some of the steps from [this guide](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc)
Here are the main steps I had to follow:


1. Create the subnetwork and secondary IP ranges in the host project that you intend to use for the GKE cluster
2. When you attach the service project during shared VPC setup, check the box for "Kubernetes engine access"
3. Run this command in the service project to make sure the subnet and secondary ranges are usable
```
gcloud container subnets list-usable \                 
--project <service project ID> \  
--network-project <Host Project ID>
```
4. Then update the cluster.jinja file created by "kfctl generate" to specify the network and subnetwork with their self links

```
cluster:
  name: {{ CLUSTER_NAME }}
  network: projects/<host project ID>/global/networks/<network name>
  subnetwork: projects/<host project ID>/regions/<region>/subnetworks/<subnet name>
  initialClusterVersion: "{{ properties['cluster-version'] }}"
```
5. In cluster.jinja, disable subnetwork creation and specify secondary IP ranges by name (ipAllocationPolicy section may have to be moved out of the IF block if you aren't setting private cluster = true)
```
{% if properties['securityConfig']['privatecluster'] %}
ipAllocationPolicy:
  createSubnetwork: false
  useIpAliases: true
  clusterSecondaryRangeName: <name of secondary ip range for pods>
  servicesSecondaryRangeName: <name of secondary ip range for services>
```
After the above changes I was able to deploy with "kfctl apply" as expected.

The [GKE API docs](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters) are also helpful if further customization to cluster.jinja is needed.

**Comment by syd-quant** [comment]
Hi @jrabiah @jlewi  

I was following above mentioned steps to setup Kubeflow on SharedVPC on GCP. 

1. Created the subnetwork and secondary IP ranges in the host project 
```
$ gcloud container subnets list-usable --project project-kubeflowsetup2 --network-project project-kubeflowsetup --format json

[
  {
    "ipCidrRange": "10.0.4.0/22",
    "network": "projects/project-kubeflowsetup/global/networks/p1-shared-vpc",
    "secondaryIpRanges": [
      {
        "ipCidrRange": "10.0.32.0/20",
        "rangeName": "p1-kf-services",
        "status": "UNUSED"
      },
      {
        "ipCidrRange": "10.4.0.0/14",
        "rangeName": "p1-kf-pods",
        "status": "UNUSED"
      }
    ],
    "subnetwork": "projects/project-kubeflowsetup/regions/us-east1/subnetworks/p1-share-subnet1"
  }
]

```
2. Updated Network in **cluster.jinja** file 

```
cluster:
      name: {{ CLUSTER_NAME }}
      network: projects/project-kubeflowsetup/global/networks/p1-shared-vpc
      subnetwork: projects/project-kubeflowsetup/regions/us-east1/subnetworks/p1-share-subnet1
      initialClusterVersion: "{{ properties['cluster-version'] }}"
```

 3. Updated Secondary IP ranges Name in **cluster.jinja** file  

```
 {% if properties['securityConfig']['privatecluster'] %}
      ipAllocationPolicy:
        createSubnetwork: false
        useIpAliases: true
        clusterSecondaryRangeName: "p1-kf-pods"
        servicesSecondaryRangeName: "p1-kf-services"
      privateClusterConfig:
```
4. Content of **cluster-kubeflow.yaml** file 

```
imports:
- path: cluster.jinja
resources:
- name: kubeflow
  properties:
    autoprovisioning-config:
      enabled: true
      max-accelerator:
      - count: 8
        type: nvidia-tesla-k80
      max-cpu: 20
      max-memory: 200
    cluster-version: "1.13.10-gke.0"
    cpu-pool-enable-autoscaling: true
    cpu-pool-initialNodeCount: 2
    cpu-pool-machine-type: n1-standard-8
    cpu-pool-max-nodes: 10
    cpu-pool-min-nodes: 0
    enable_tpu: true
    gkeApiVersion: v1beta1
    gpu-number-per-node: 1
    gpu-pool-enable-autoscaling: true
    gpu-pool-initialNodeCount: 0
    gpu-pool-machine-type: n1-standard-8
    gpu-pool-max-nodes: 0
    gpu-pool-min-nodes: 0
    gpu-type: nvidia-tesla-k80
    ipName: kf101-ip
    pool-version: v1
    securityConfig:
      masterAuthorizedNetworksConfigCidr:
      - cidrBlock: 1.2.3.4/32
      masterAuthorizedNetworksConfigEnabled: false
      masterIpv4CidrBlock: 172.16.0.16/28
      podSecurityPolicy: false
      privatecluster: false
      secureNodeMetadata: false
    users:
    - user:user@example.com
    zone: us-east1-b
  type: cluster.jinja
```

5. Getting following error while executing `kfctl apply all -V` command: 

> ERRO[0027] Updating kubeflow-kf error: &{Code:RESOURCE_ERROR Location:/deployments/kubeflow-kf/resources/kubeflow-kf Message:{"ResourceType":"gcp-types/container-v1beta1:projects.locations.clust
> ers","ResourceErrorCode":"400","ResourceErrorMessage":{"code":400,"message":"Names of existing secondary ranges must be specified when using a shared network.","status":"INVALID_ARGU
> MENT","statusMessage":"Bad Request","requestPath":"https://container.googleapis.com/v1beta1/projects/project-kubeflowsetup2/locations/us-east1-b/clusters","httpMethod":"POST"}} ForceSen
> dFields:[] NullFields:[]}  filename="gcp/gcp.go:348"
> Error: couldn't apply KfApp:  (kubeflow.error): Code 500 with message: coordinator Apply failed for gcp:  (kubeflow.error): Code 400 with message: gcp apply could not update deployme
> nt manager Error could not update deployment manager entries; Updating kubeflow-kf error(400): BAD REQUEST 

I have mentioned the Names of existing secondary ranges but still getting the error. Is there anything I am missing.

**Comment by jrabiah** [comment]
@syd-quant 
From the cluster-kubeflow.yaml, it looks like "privatecluster" is set to false. If you look back to cluster.jinja, the ipAllocationPolicy sections which specifies the secondary ranges comes after an 'if' statement which checks the value of "privatecluster". I suspect this section is just being skipped over and would try moving the ipAllocationPolicy items outside of the 'if' block.

**Comment by syd-quant** [comment]
Thanks @jrabiah It worked!

**Comment by jlewi** [comment]
kubeflow/kubeflow#3210 to bulk replace the images.

Right now `kfctl set image-name` emits a list of images in the form `${OLD}=${NEW}` to stdout. This could be used as the basis to build a GCB workflow to replicate the images.

**Comment by jlewi** [comment]
The instructions: https://www.kubeflow.org/docs/gke/private-clusters/

Need to be updated.

We should put a warning on that page that the instructions are outdated until its been fixed.

Here are some of the changes that are needed

1. The instructions for replicating images to private GCR need to be fixed 

    * A good solution is blocked on tooling to replicate the images (#3210 

1. Step 7 removing cert-manager is no longer applicable with 0.7 because we are using managed gke certificates which should work with private GKE

    * We should be able to remove step 10 as well since you should no longer need to upload a certificate since you can use managed certificates

**Comment by amygdala** [comment]
noooo, stale[bot]. (I assume we want to keep this one open)

**Comment by jlewi** [comment]
It looks like I might have muddied the waters on this issue.  

* The original issue was about [shared VPC](https://cloud.google.com/vpc/docs/shared-vpc)
* At some point I started conflating that with [VPC Service Controls](https://cloud.google.com/vpc-service-controls/docs/)

Lets use this issue to track documenting shared VPC and I'll open a new issue for VPC service controls

**Comment by damitkwr** [comment]
I wanted to add another parameter to  @syd-quant that I encountered. As per the Google documentation [here](https://cloud.google.com/kubernetes-engine/docs/how-to/flexible-pod-cidr) , I ran into a problem where my subnet was only allocated a /24 CIDR  secondary ranges so when the deployment was happening I ran into the problem where IP range was exhausted for the secondary ranges.

To fix this problem, we have to limit the number of maximum pods per nodes so it doesn't exhaust the IP space. To do this add this line according to whatever limit you have on your **cluster.jinja**:

>       defaultMaxPodsConstraint:
>         maxPodsPerNode: 64

You should be good to go now.

**Comment by DesmondH0** [comment]
Just an update on this thread
I just followed @jrabiah steps to setup the kubeflow in shared VPC
And 1 thing was missing is we need to grant the `Compute Network Admin` to `${CLUSTER}-admin@clover-non-production.iam.gserviceaccount.com` in target VPC project

Otherwise istio would stuck with `origin authentication failed` and never pass the health check in GCP LB

**Comment by jlewi** [comment]
@DesmondH0 , @jrabiah, @damitkwr @syd-quant  Is anyone interested in collecting the information in this issue into a PR providing a doc either in
https://github.com/kubeflow/gcp-blueprints

Or on the website
https://github.com/kubeflow/website

**Comment by thesuperzapper** [comment]
This issue should live on the GCP distribution repo: https://github.com/kubeflow/gcp-blueprints

Please raise it there if this is still an issue.

/close
